crates/samples/dcomp/src/main.rs

```diff
@@ -0,0 +1,781 @@
+                CW_USEDEFAULT,
+                CW_USEDEFAULT,
+                CW_USEDEFAULT,
+                CW_USEDEFAULT,
+                None,
+                None,
+                instance,
+                Some(self as *mut _ as _),
+            );
+
+            debug_assert!(handle.0 != 0);
+            debug_assert!(handle == self.handle);
+            let mut message = MSG::default();
+
+            while GetMessageA(&mut message, HWND(0), 0, 0).into() {
+                DispatchMessageA(&message);
+            }
+
+            Ok(())
+        }
+    }
+
+    extern "system" fn wndproc(
+        window: HWND,
+        message: u32,
+        wparam: WPARAM,
+        lparam: LPARAM,
+    ) -> LRESULT {
+        unsafe {
+            if message == WM_NCCREATE {
+                let cs = lparam.0 as *const CREATESTRUCTA;
+                let this = (*cs).lpCreateParams as *mut Self;
+                (*this).handle = window;
+
+                SetWindowLongPtrA(window, GWLP_USERDATA, this as _);
+            } else {
+                let this = GetWindowLongPtrA(window, GWLP_USERDATA) as *mut Self;
+
+                if !this.is_null() {
+                    return (*this).message_handler(message, wparam, lparam);
+                }
+            }
+
+            DefWindowProcA(window, message, wparam, lparam)
+        }
+    }
+}
+
+fn create_text_format() -> Result<IDWriteTextFormat> {
+    unsafe {
+        let factory: IDWriteFactory2 = DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED)?;
+
+        let format = factory.CreateTextFormat(
+            w!("Candara"),
+            None,
+            DWRITE_FONT_WEIGHT_NORMAL,
+            DWRITE_FONT_STYLE_NORMAL,
+            DWRITE_FONT_STRETCH_NORMAL,
+            CARD_HEIGHT / 2.0,
+            w!("en"),
+        )?;
+
+        format.SetTextAlignment(DWRITE_TEXT_ALIGNMENT_CENTER)?;
+        format.SetParagraphAlignment(DWRITE_PARAGRAPH_ALIGNMENT_CENTER)?;
+        Ok(format)
+    }
+}
+
+fn create_image() -> Result<IWICFormatConverter> {
+    unsafe {
+        let factory: IWICImagingFactory2 =
+            CoCreateInstance(&CLSID_WICImagingFactory, None, CLSCTX_INPROC_SERVER)?;
+
+        let decoder = factory.CreateDecoderFromFilename(
+            w!("image.jpg"),
+            std::ptr::null(),
+            GENERIC_READ,
+            WICDecodeMetadataCacheOnDemand,
+        )?;
+
+        let source = decoder.GetFrame(0)?;
+        let image = factory.CreateFormatConverter()?;
+
+        image.Initialize(
+            &source,
+            &GUID_WICPixelFormat32bppBGR,
+            WICBitmapDitherTypeNone,
+            None,
+            0.0,
+            WICBitmapPaletteTypeMedianCut,
+        )?;
+
+        Ok(image)
+    }
+}
+
+fn create_device_3d() -> Result<ID3D11Device> {
+    let mut flags = D3D11_CREATE_DEVICE_BGRA_SUPPORT;
+
+    if cfg!(debug_assertions) {
+        flags |= D3D11_CREATE_DEVICE_DEBUG;
+    }
```

Comment:
I'm not sure where you get the debug layers these days; this doesn't work on my retail Windows 11 machines.

```
Error { code: HRESULT(0x887A002D), message: "The application requested an operation that depends on an SDK component that is missing or mismatched."
```

Reply:
Update: We can add this to the `README.md`:

To run this sample, you will need the Direct3D 11 debug layer installed. To install the layer, go to Settings > Apps > Optional features and add the optional feature `Graphics Tools`.